### ISSUE TITLE
Cannot prophesize and interface extending \Throwable

### ISSUE THREAD

**Issue by dunglas** [phpspec/prophecy#410]
If an exception interface extends `\Throwable` (now a recommended best practice, enforced by phpstan), it's not possible to prophesize it.

Reproducer:

```php
interface MyException extends \Throwable {}

$e = $this->prophesize(MyException::class);
$e->willExtend(\Exception::class);
$e->reveal();
```

(moved from https://github.com/symfony/symfony/issues/27419)

**Comment by ciaranmcnulty** [comment]
What error do you get?

**Comment by dunglas** [comment]
> Prophecy\Exception\Doubler\MethodNotExtendableException: Method `getMessage` is not extendable, so can not be added.

**Comment by Taluu** [comment]
Should be closed @ciaranmcnulty, as you fixed it in #412

**Comment by ostrolucky** [comment]
Yes and included test does not use same thing code posted here does. It does not extend Exception but stdClass. So once again, you have failing test case in this issue, which was never fixed.

**Comment by ciaranmcnulty** [comment]
I’m not clear why the example adds willExtend(Exception) - can you clarify why the test is adding that?

**Comment by ostrolucky** [comment]
Actually, patch explicitly forbids this use case with test case in https://github.com/phpspec/prophecy/blob/3660b2eed90abdf92ac2e9dae9a238d2701bfa77/spec/Prophecy/Doubler/ClassPatch/ThrowablePatchSpec.php#L42-L51

**Comment by ciaranmcnulty** [comment]
No that is excluding concrete Exceptions or Errors (which don’t need patching)

**Comment by ostrolucky** [comment]
> can you clarify why the test is adding that?

PHP5 compatible library can't use `->willImplement('Throwable')` so has to do `->willExtend('Exception')` in case of mocking interface known to be thrown

```php
$mayOrMayNotBeThrowable = rand(0,1) ? 'Throwable' : 'stdClass';
$this->prophesize($mayOrMayNotBeThrowableClass)->willExtend(\Exception::class)->reveal();
```

**Comment by dunglas** [comment]
For those wanting looking for a quick userland fix supporting the use case described  by @ostrolucky (basically, Symfony), this works:

```php
$serializerException = $this->prophesize(ExceptionInterface::class);
if (!is_a(ExceptionInterface::class, \Throwable::class, true)) {
    $serializerException->willExtend(\Exception::class);
}
```

**Comment by ostrolucky** [comment]
@dunglas You don't support PHP5 so in your case you can just use `->willImplement('Throwable')`, can't you?

**Comment by dunglas** [comment]
Yes it will work too (but we'll be less accurate).

**Comment by ciaranmcnulty** [comment]
I’m confused because the case in @dunglas workaround looks the same as the case in the issue

Also - are we talking about a case where Throwable has been defined by some BC shim?

**Comment by ostrolucky** [comment]
It's not 1:1 replacement, but workaround. Goal is to be able to pass typehints for both `interface MyException implements Throwable` and `Exception`. Prophecy should generate something like `class MockClass extends Exception implements MyException`, not sure why does it give up as soon as it sees there is one final method in `Exception`

**Comment by TiMESPLiNTER** [comment]
```
public function it_throws_exception(
        \Throwable $exception
    ): void {
        $exceptionFile = 'file.php';
        $exception->getFile()->willReturn($exceptionFile);
}
```

leads to:

```
- it throws exception on sync invoices payment status if producing message fails
      can not add prophecy for a method `Double\Throwable\P25::getFile()`
      as it is a final method.
```

with phpspec 6.2.1 although there's no `final` definition in the interface?

**Comment by ciaranmcnulty** [comment]
@TiMESPLiNTER unfortunately because Throwable is not implementable in user land, when you ask for a Throwable we have to extend either Exception or Error for the double.

Both of those types have `getFile` as final so I can't see a way to do what you're after
